bin/akula-rpc.rs

```diff
@@ -43,15 +124,572 @@ where
+        let block_hash = canonical_hash::read(&self.db.begin().await?, block_number)
+            .await?
+            .unwrap();
+
+        let header = header::read(&self.db.begin().await?, block_hash, block_number)
+            .await?
+            .unwrap();
+
+        let tx = &self.db.begin().await?;
+        let mut state = Buffer::new(tx, BlockNumber(0), Some(block_number));
+        let mut analysis_cache = AnalysisCache::default();
+        let block_spec = MAINNET.collect_block_spec(block_number);
+
+        let value = match call_data.value {
+            None => Default::default(),
+            Some(v) => v,
+        };
+
+        let input = match call_data.data {
+            None => Default::default(),
+            Some(i) => i,
+        };
+
+        let sender = match call_data.from {
+            None => Default::default(),
+            Some(s) => s,
+        };
+
+        let msg_with_sender = MessageWithSender {
+            message: Message::Legacy {
+                chain_id: Some(ChainId(1)),
+                nonce: 0,
+                gas_price: Default::default(),
+                gas_limit: 0,
+                action: TransactionAction::Call(call_data.to),
+                value,
+                input,
+            },
+            sender,
+        };
+
+        let gas = match call_data.gas {
+            None => Default::default(),
+            Some(g) => g,
+        };
+
+        Ok(execute(
+            &mut IntraBlockState::new(&mut state),
+            None,
+            &mut analysis_cache,
+            &PartialHeader::from(header.clone()),
+            &block_spec,
+            &msg_with_sender,
+            gas.0[0],
+        )
+        .await?
+        .output_data)
+    }
+
+    async fn estimate_gas(&self, call_data: CallData, block_number: BlockNumber) -> RpcResult<U64> {
+        let block_hash = canonical_hash::read(&self.db.begin().await?, block_number)
+            .await?
+            .unwrap();
+
+        let header = header::read(&self.db.begin().await?, block_hash, block_number)
+            .await?
+            .unwrap();
+
+        let nonce = account::read(
+            &self.db.begin().await?,
+            call_data.from.unwrap(),
+            Some(block_number),
+        )
+        .await?
+        .map(|acc| acc.nonce)
+        .unwrap();
+
+        let value = match call_data.value {
+            None => Default::default(),
+            Some(v) => v,
+        };
+
+        let input = match call_data.data {
+            None => Default::default(),
+            Some(i) => i,
+        };
+
+        let sender = match call_data.from {
+            None => Default::default(),
+            Some(s) => s,
+        };
+
+        let msg = MessageWithSender {
+            message: Message::Legacy {
+                chain_id: Some(ChainId(1)),
+                nonce,
+                gas_price: Default::default(),
+                gas_limit: header.gas_limit,
```

Comment:
How can I get a pending block? [Actually gas_limit should be set to the value of the gas field in CallData or the block gas limit of the pending block](https://eth.wiki/json-rpc/API#eth_estimategas).

Reply:
I'm not sure if we have this concept yet. For now you could return an error in case if the user doesn't specify gas_limit (so make it mandatory for now).

Reply:
Alright, thanks!